File: backend/ecc.py
import hashlib

SHA3_PREFIX_LEN = 8

class ModNum:
    def __init__(self, val: int, mod: int):
        assert isinstance(val, int)
        assert isinstance(mod, int)
        self.val = val % mod
        self.mod = mod

    def modArithAssert(self, other):
        assert isinstance(other, ModNum)
        assert self.mod == other.mod

    def modArithConvert(self, other):
        if isinstance(other, int):
            other = ModNum(other, self.mod)
        elif isinstance(other, ModNum):
            pass
        return other

    def __add__(self, other):
        other = self.modArithConvert(other)
        self.modArithAssert(other)
        return ModNum(self.val + other.val, self.mod)

    def __radd__(self, other):
        return self + other

    def __sub__(self, other):
        other = self.modArithConvert(other)
        self.modArithAssert(other)
        return ModNum(self.val + self.mod - other.val, self.mod)

    def __mul__(self, other):
        other = self.modArithConvert(other)
        self.modArithAssert(other)
        return ModNum(self.val * other.val, self.mod)

    def __rmul__(self, other):
        return self * other

    def __rfloordiv__(self, other):
        if isinstance(other, int):
            return ModNum(other, self.mod) // self
        else:
            raise TypeError(f'Unsupported // between {type(other)} and ModNum')

    def __floordiv__(self, other):
        other = self.modArithConvert(other)
        self.modArithAssert(other)
        x = gcdExtend(other.val, self.mod)

        return self * ModNum(x, self.mod)

    def __repr__(self):
        return f'{self.val} (%{self.mod})'

    def __pow__(self, exp):
        assert isinstance(exp, int) and exp > 0

        if exp == 1:
            return ModNum(self.val, self.mod)
        sq = self ** (exp // 2)
        sq *= sq
        if exp % 2:
            return sq * self
        else:
            return sq

    def __gt__(self, other):
        if isinstance(other, int):
            return self.val > other
        elif isinstance(other, ModNum):
            return self.val > other.val
        else:
            raise TypeError(f'gt between ModNum and {type(other)} not supported')

    def __eq__(self, other):
        if isinstance(other, int):
            return self.val == other
        elif isinstance(other, ModNum):
            return self.val == other.val and self.mod == other.mod
        else:
            raise TypeError(f'eq between ModNum and {type(other)} not supported')

    def __ge__(self, other):
        return self == other or self > other
    
    def __lt__(self, other):
        if isinstance(other, int):
            return self.val < other
        elif isinstance(other, ModNum):
            return self.val < other.val
        else:
            raise TypeError(f'lt between ModNum and {type(other)} not supported')

    def __le__(self, other):
        return self == other or self < other


def gcdExtend(ppr: int, pr: int):
    m = pr
    ppx = 1
    px = 0
    while pr != 1:
        q = ppr // pr
        r = ppr % pr
        x = (ppx - q * px) % m

        ppr, pr = pr, r
        ppx, px = px, x
    return px


class ECurve:
    """
    Defines the montgomery curve: y^2 = x^3 + Ax + B
    """
    def __init__(self, A, B):
        assert isinstance(A, int)
        assert isinstance(B, int)
        self.A = A
        self.B = B

    def __eq__(self, other):
        assert isinstance(other, ECurve)
        return self.A == other.A and self.B == other.B


class EPoint:
    x: ModNum
    y: ModNum
    curve: ECurve

    def __init__(self, x: ModNum, y: ModNum, curve: ECurve):
        assert isinstance(x, ModNum)
        assert isinstance(y, ModNum)
        assert x.mod == y.mod
        assert isinstance(curve, ECurve)
        self.x = x
        self.y = y
        self.curve = curve

    def pointArithAssert(self, other):
        assert isinstance(other, EPoint)
        assert self.curve == other.curve
        assert self.x.mod == other.x.mod

    def __add__(self, other):
        self.pointArithAssert(other)
        if self == other:
            return self.double()
        l = (other.y - self.y) // (other.x - self.x)
        return self.intersect(other, l)
        
    def intersect(self, other, l: ModNum):
        x = l * l - self.x - other.x
        y = l * (self.x - x) - self.y
        return EPoint(x, y, self.curve)

    def double(self):
        l = (3 * self.x ** 2 + self.curve.A) // (2 * self.y)
        return self.intersect(self, l)

    def __mul__(self, times):
        assert isinstance(times, int) and times > 0
        if times == 1:
            return EPoint(ModNum(self.x.val, self.x.mod), ModNum(self.y.val, self.y.mod), self.curve)
        sq = self * (times // 2)
        sq = sq.double()
        if times % 2:
            return self + sq
        else:
            return sq

    def __rmul__(self, times):
        return self * times

    def __eq__(self, other):
        assert isinstance(other, EPoint)
        return self.x == other.x and self.y == other.y and self.curve == other.curve
    
    def __repr__(self):
        return f'({self.x.val}, {self.y.val})'


def mysecp():
    A = 0x5e924cd447a56b
    B = 0x892f0a953f589b
    curve = ECurve(A, B)
    mod = 0xbcffb098340493
    x = ModNum(0x9a77dc33b36acc, mod)
    y = ModNum(0x279be90a95dbdd, mod)
    assert y * y == x * x * x + A * x + B
    G = EPoint(x, y, curve)
    return curve, G, 0xbcffb09c43733d


class ECC:
    def __init__(self, curve: ECurve, G: EPoint, order: int, d: int):
        self.curve = curve
        self.G = G
        self.order = order
        self.d = d

    def pub(self) -> EPoint:
        return self.d * self.G

    def sign(self, message: bytes):
        z = int.from_bytes(hashlib.sha3_256(message).digest()[:SHA3_PREFIX_LEN], 'little')
        n = self.order
        r, s = ModNum(0, n), ModNum(0, n)
        while s == 0:
            while r == 0:
                k = 2
                P = k * self.G
                r = ModNum(P.x.val, n)
            s = (z + r * self.d) // k
        return r, s

    def verify(self, message: bytes, pub: EPoint, r: ModNum, s: ModNum):
        n = self.order
        if not 1 <= r <= n-1 or not 1 <= s <= n-1:
            return False
        z = ModNum(int.from_bytes(hashlib.sha3_256(message).digest()[:SHA3_PREFIX_LEN], 'little'), r.mod)
        u1: ModNum = z // s
        u2: ModNum = r // s
        P: EPoint = u1.val * self.G + u2.val * pub
        return P.x.val == r.val

File: sw/provisioner/src/ecc.rs
//! Private-key generation for badge provisioning.
//!
//! Port of pcb-util/dev/ecc.py `gen_key`: a private key is a random scalar `d`
//! on a custom ~56-bit short-Weierstrass curve, retried until the public key
//! `d*G` has the requested y-parity. Because the field is only ~56 bits every
//! product fits in u128, so no bignum library is needed.

use rand::RngExt;

// Curve y^2 = x^3 + A*x + B (mod P); generator G, group order N.
// Values match pcb-util/dev/ecc.py `mysecp`.
const A: u128 = 0x5e924cd447a56b;
// Curve coefficient B: only the on-curve test references it; the add/double formulas need A only.
#[allow(dead_code)]
const B: u128 = 0x892f0a953f589b;
const P: u128 = 0xbcffb098340493;
const N: u128 = 0xbcffb09c43733d;
const GX: u128 = 0x9a77dc33b36acc;
const GY: u128 = 0x279be90a95dbdd;
const G: Point = Point { x: GX, y: GY };

/// Private key length in bytes: ceil(bit_length(N) / 8) = ceil(56 / 8) = 7.
pub(crate) const PRIV_KEY_LEN: usize = 7;

// --- Field arithmetic mod P (mirrors ecc.py ModNum) ---

fn add(a: u128, b: u128) -> u128 {
    (a + b) % P
}

fn sub(a: u128, b: u128) -> u128 {
    // a + P - b avoids unsigned underflow; matches ModNum.__sub__.
    (a + P - b) % P
}

fn mul(a: u128, b: u128) -> u128 {
    // a, b < P < 2^56, so a*b < 2^112 fits in u128.
    (a * b) % P
}

/// Modular inverse of `a` mod P via the extended Euclidean algorithm.
/// Faithful port of ecc.py `gcdExtend(a, P)`; requires `a` coprime to P (P prime, a != 0).
fn inv(a: u128) -> u128 {
    let m = P as i128;
    let mut ppr = a as i128;
    let mut pr = P as i128;
    let mut ppx: i128 = 1;
    let mut px: i128 = 0;
    while pr != 1 {
        let q = ppr / pr;
        let r = ppr % pr;
        let x = ((ppx - q * px) % m + m) % m; // keep non-negative like Python's %
        ppr = pr;
        pr = r;
        ppx = px;
        px = x;
    }
    px as u128
}

/// a / b mod P (ModNum.__floordiv__).
fn div(a: u128, b: u128) -> u128 {
    mul(a, inv(b))
}

// --- Curve point arithmetic (affine; mirrors ecc.py EPoint) ---

#[derive(Clone, Copy, PartialEq, Eq)]
struct Point {
    x: u128,
    y: u128,
}

/// Third intersection point given slope l (ecc.py EPoint.intersect).
fn intersect(p: Point, q: Point, l: u128) -> Point {
    let x = sub(sub(mul(l, l), p.x), q.x);
    let y = sub(mul(l, sub(p.x, x)), p.y);
    Point { x, y }
}

/// Point doubling (ecc.py EPoint.double): l = (3*x^2 + A) / (2*y).
fn double(p: Point) -> Point {
    let l = div(add(mul(3, mul(p.x, p.x)), A), mul(2, p.y));
    intersect(p, p, l)
}

/// Point addition (ecc.py EPoint.__add__): doubles when the points are equal.
fn point_add(p: Point, q: Point) -> Point {
    if p == q {
        return double(p);
    }
    let l = div(sub(q.y, p.y), sub(q.x, p.x));
    intersect(p, q, l)
}

/// Scalar multiplication `times * p` by double-and-add.
/// Faithful port of ecc.py EPoint.__mul__ (recursive); requires `times >= 1`.
fn scalar_mul(times: u128, p: Point) -> Point {
    if times == 1 {
        return p;
    }
    let sq = double(scalar_mul(times / 2, p));
    if times % 2 == 1 { point_add(p, sq) } else { sq }
}

/// Generate a private key for the given team parity (BLUE = 0, RED = 1).
/// Retries fresh random scalars until the public key `d*G` has y-parity == `parity`.
/// Returns `d` as 7 bytes little-endian (matches ecc.py `gen_key`).
pub(crate) fn gen_key(parity: u8) -> [u8; PRIV_KEY_LEN] {
    assert!(parity == 0 || parity == 1, "parity must be 0 or 1");
    let mut rng = rand::rng();
    loop {
        let d = rng.random_range(1..N); // d in [1, N-1]
        let pubkey = scalar_mul(d, G);
        if (pubkey.y & 1) as u8 == parity {
            let le = d.to_le_bytes();
            let mut out = [0u8; PRIV_KEY_LEN];
            out.copy_from_slice(&le[..PRIV_KEY_LEN]);
            return out;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Whether a point satisfies y^2 == x^3 + A*x + B (mod P).
    fn on_curve(p: Point) -> bool {
        let lhs = mul(p.y, p.y);
        let rhs = add(add(mul(mul(p.x, p.x), p.x), mul(A, p.x)), B);
        lhs == rhs
    }

    // Reference vectors for d*G, computed with pcb-util/dev/ecc.py: (d, x, y).
    const VECTORS: &[(u128, u128, u128)] = &[
        (0x1, 0x9a77dc33b36acc, 0x279be90a95dbdd),
        (0x2, 0x9a2e8af9f0403c, 0x52b00e5f0b8ee1),
        (0x3, 0x2415048f5fa910, 0x2ae9d459887b45),
        (0x7, 0x12fd78df99ee27, 0x446b65155e5d77),
        (0x1234567890abcd, 0x27c9ab19481936, 0x894ad3c33393b1),
        (0xbcffb09c43733c, 0x9a77dc33b36acc, 0x9563c78d9e28b6), // (N-1)*G = -G
        (0xb3a73ce2ff2, 0x9c0795e840b788, 0x52bcfd7eb57491),
    ];

    #[test]
    fn generator_on_curve() {
        assert!(on_curve(G));
    }

    #[test]
    fn scalar_mul_matches_python() {
        for &(d, x, y) in VECTORS {
            let pt = scalar_mul(d, G);
            assert_eq!(pt.x, x, "x mismatch for d={d:#x}");
            assert_eq!(pt.y, y, "y mismatch for d={d:#x}");
            assert!(on_curve(pt), "off curve for d={d:#x}");
        }
    }

    #[test]
    fn priv_key_encoding_is_le7() {
        // d.to_le_bytes()[..7] must equal Python d.to_bytes(7, 'little').
        let d: u128 = 0x1234567890abcd;
        assert_eq!(
            &d.to_le_bytes()[..PRIV_KEY_LEN],
            &[0xcd, 0xab, 0x90, 0x78, 0x56, 0x34, 0x12]
        );
    }

    #[test]
    fn gen_key_respects_parity_and_range() {
        for parity in [0u8, 1u8] {
            let key = gen_key(parity);
            // Reconstruct d and confirm d*G has the requested parity.
            let mut buf = [0u8; 16];
            buf[..PRIV_KEY_LEN].copy_from_slice(&key);
            let d = u128::from_le_bytes(buf);
            assert!(d >= 1 && d < N, "d out of range for parity={parity}");
            let pubkey = scalar_mul(d, G);
            assert_eq!((pubkey.y & 1) as u8, parity, "parity mismatch");
        }
    }
}
